pkg/intermediate/aggregate.go

```diff
@@ -0,0 +1,284 @@
+	} else if workerNum <= 0 {
+		return nil, fmt.Errorf("Worker number cannot be <= 0.")
+	}
+	return &aggregation{
+		make(map[Tuple][]entities.Record),
+		sync.RWMutex{},
+		make(chan chan *entities.Message),
+		messageChan,
+		workerNum,
+		make([]*worker, 0),
+		correlateFields,
+	}, nil
+}
+
+func (a *aggregation) Start() {
+	for i := 0; i < a.workerNum; i++ {
+		w := createWorker(i, a.workerPool, a.AggregateMsgBy5Tuple)
+		w.start()
+		a.workerList = append(a.workerList, w)
+	}
+	for message := range a.messageChan {
+		channel := <-a.workerPool
+		channel <- message
+	}
+}
+
+func (a *aggregation) Stop() {
+	for _, worker := range a.workerList {
+		worker.stop()
+	}
+}
+
+// AggregateMsgBy5Tuple gets 5-tuple info from records in message and stores in cache
+func (a *aggregation) AggregateMsgBy5Tuple(message *entities.Message) error {
+	addOriginalExporterInfo(message)
+	if message.Set.GetSetType() == entities.Template { // skip template records
+		return nil
+	}
+	records := message.Set.GetRecords()
+	for _, record := range records {
+		tuple, err := getTupleFromRecord(record)
+		if err != nil {
+			return err
+		}
+		a.correlateRecords(tuple, record)
+	}
+	return nil
+}
+
+func (a *aggregation) GetTupleRecordMap() map[Tuple][]entities.Record {
+	a.tupleRecordLock.RLock()
+	defer a.tupleRecordLock.RUnlock()
+	return a.tupleRecordMap
+}
+
+func (a *aggregation) DeleteTupleFromMap(tuple Tuple) {
+	a.tupleRecordLock.Lock()
+	defer a.tupleRecordLock.Unlock()
+	delete(a.tupleRecordMap, tuple)
+}
+
+// correlateRecords fills records info by correlating incoming and current records
+func (a *aggregation) correlateRecords(tuple Tuple, record entities.Record) {
+	existingRecords := a.GetTupleRecordMap()[tuple]
+	// only fill the information for record from source node
+	if isRecordFromSrc(record) {
+		var isFilled bool
+		for _, existingRec := range existingRecords {
+			for _, field := range a.correlateFields {
+				if containsInfoElement(record.GetInfoElementMap(), field) {
+					record.GetInfoElementMap()[field].Value = existingRec.GetInfoElementMap()[field].Value
+					isFilled = true
+				}
+			}
+			if isFilled {
+				break
+			}
+		}
+	} else {
+		for _, existingRec := range existingRecords {
+			if isRecordFromSrc(existingRec) {
+				for _, field := range a.correlateFields {
+					if containsInfoElement(record.GetInfoElementMap(), field) {
+						existingRec.GetInfoElementMap()[field].Value = record.GetInfoElementMap()[field].Value
+					}
+				}
+			}
+		}
+	}
+	a.addRecordToMap(tuple, record)
+	a.removeDuplicates(tuple)
+}
+
+func (a *aggregation) removeDuplicates(tuple Tuple) {
+	a.tupleRecordLock.Lock()
+	defer a.tupleRecordLock.Unlock()
+	records := a.tupleRecordMap[tuple]
+	srcRecords := make([]entities.Record, 0)
+	dstRecords := make([]entities.Record, 0)
+	for _, record := range records {
+		if isRecordFromSrc(record) {
+			srcRecords = append(srcRecords, record)
+		} else {
```

Comment:
Do we need to check destinationPodName existing or not?

Reply:
the correlate the assumption for using `isRecordFromSrc` is that the record will have both sourcePodName and destinationPodName (specific to deal with records exported from Antrea flow exporter)